packages/credit-card-integration/src/CreditCardPaymentMethodComponent.tsx

```diff
@@ -328,82 +112,159 @@ class CreditCardPaymentMethodComponent extends Component<
-            statuses: { isLoadingInstruments },
-        } = checkoutState;
+    const isInitialRenderRef = useRef(true);
 
-        const config = getConfig();
-        const customer = getCustomer();
+    useEffect(() => {
+        if (isInitialRenderRef.current) {
+            isInitialRenderRef.current = false;
 
-        if (!config || !customer || !method) {
-            throw new Error('Unable to get checkout');
+            return;
         }
 
-        const instruments = this.filterInstruments(getInstruments(method));
-        const isInstrumentFeatureAvailableProp = isInstrumentFeatureAvailable({
-            config,
-            customer,
-            isUsingMultiShipping,
-            paymentMethod: method,
-        });
-
-        return {
-            instruments,
-            isCardCodeRequired: method.config.cardCode || method.config.cardCode === null,
-            isCustomerCodeRequired: !!method.config.requireCustomerCode,
-            isInstrumentCardCodeRequired: isInstrumentCardCodeRequiredSelector(checkoutState),
-            isInstrumentCardNumberRequired: isInstrumentCardNumberRequiredSelector(checkoutState),
-            isInstrumentFeatureAvailable: isInstrumentFeatureAvailableProp,
-            isLoadingInstruments: isLoadingInstruments(),
-            isPaymentDataRequired: isPaymentDataRequired(),
-            loadInstruments: checkoutService.loadInstruments,
-            shouldShowInstrumentFieldset:
-                isInstrumentFeatureAvailableProp && instruments.length > 0,
+        const reInit = async () => {
+            try {
+                setValidationSchema(method, getValidationSchema());
+
+                await deinitializePayment({
+                    gatewayId: method.gateway,
+                    methodId: method.id,
+                });
+
+                await initializePayment(
+                    {
+                        gatewayId: method.gateway,
+                        methodId: method.id,
+                    },
+                    selectedInstrument,
+                );
+            } catch (error: unknown) {
+                if (error instanceof Error) {
+                    onUnhandledError(error);
+                }
+            }
         };
-    }
-}
 
-export default CreditCardPaymentMethodComponent;
+        void reInit();
+    }, [method, selectedInstrumentId, isAddingNewCard]);
+
+    useEffect(() => {
+        return () => {
+            const deInit = async () => {
+                try {
+                    setValidationSchema(method, null);
+
+                    await deinitializePayment({
+                        gatewayId: method.gateway,
+                        methodId: method.id,
+                    });
+                } catch (error) {
+                    if (error instanceof Error) {
+                        onUnhandledError(error);
+                    }
+                }
+            };
+
+            void deInit();
```

Comment:
I have created `init()`, `reInit()` and `deInit()` inside hooks.
Using explicit names makes their purpose clearer.
We may need to align on a consistent naming for these methods. ⏩️